superset-frontend/src/middleware/asyncEvent.ts

```diff
@@ -25,178 +24,240 @@ import {
-    };
+const removeListener = (id: string) => {
+  if (!listenersByJobId[id]) return;
+  delete listenersByJobId[id];
+};
 
-    const processEvents = async () => {
-      let queuedComponents = getPendingComponents(store.getState());
-      const eventArgs = lastReceivedEventId
-        ? { last_id: lastReceivedEventId }
-        : {};
-      const events: AsyncEvent[] = [];
-      if (queuedComponents && queuedComponents.length) {
-        try {
-          const { result: events } = await fetchEvents(eventArgs);
-          // refetch queuedComponents due to race condition where results are available
-          // before component state is updated with asyncJobId
-          queuedComponents = getPendingComponents(store.getState());
-          if (events && events.length) {
-            const componentsByJobId = queuedComponents.reduce((acc, item) => {
-              acc[item.asyncJobId] = item;
-              return acc;
-            }, {});
-            const fetchDataEvents: Promise<CachedDataResponse>[] = [];
-            events.forEach((asyncEvent: AsyncEvent) => {
-              const component = componentsByJobId[asyncEvent.job_id];
-              if (!component) {
-                console.warn(
-                  'Component not found for job_id',
-                  asyncEvent.job_id,
-                );
-                return setLastId(asyncEvent);
-              }
-              const componentId = component.id;
-              switch (asyncEvent.status) {
-                case JOB_STATUS.DONE:
-                  fetchDataEvents.push(
-                    fetchCachedData(asyncEvent, componentId),
-                  );
-                  break;
-                case JOB_STATUS.ERROR:
-                  store.dispatch(
-                    errorAction(componentId, [parseErrorJson(asyncEvent)]),
-                  );
-                  break;
-                default:
-                  console.warn('Received event with status', asyncEvent.status);
-              }
-
-              return setLastId(asyncEvent);
-            });
-
-            const fetchResults = await Promise.all(fetchDataEvents);
-            fetchResults.forEach(result => {
-              const data = Array.isArray(result.data)
-                ? result.data
-                : [result.data];
-              if (result.status === 'success') {
-                store.dispatch(successAction(result.componentId, data));
-              } else {
-                store.dispatch(errorAction(result.componentId, data));
-              }
-            });
+export const waitForAsyncData = async (asyncResponse: AsyncEvent) =>
+  new Promise((resolve, reject) => {
+    const jobId = asyncResponse.job_id;
+    const listener = async (asyncEvent: AsyncEvent) => {
+      switch (asyncEvent.status) {
+        case JOB_STATUS.DONE: {
+          let { data, status } = await fetchCachedData(asyncEvent); // eslint-disable-line prefer-const
+          data = Array.isArray(data) ? data : [data];
```

Comment:
`superset-ui/core` has a convenient util for this: `data = ensureIsArray(data)`: https://github.com/apache-superset/superset-ui/blob/a3ec94f7a72a3dbb6cb6563e7ea7d3582661ea43/packages/superset-ui-core/src/utils/ensureIsArray.ts#L24-L29